KEP-5773: DRA ResourceSlice Priority (#5794)
kubernetes/enhancements: keps/sig-scheduling/5773-resourceslice-priority/README.md

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

The troubleshooting section in https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4381-dra-structured-parameters#troubleshooting
still applies.

###### How does this feature react if the API server and/or etcd is unavailable?

See https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4381-dra-structured-parameters#how-does-this-feature-react-if-the-api-server-andor-etcd-is-unavailable

###### What are other known failure modes?

See https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4381-dra-structured-parameters#what-are-other-known-failure-modes.

###### What steps should be taken if SLOs are not being met to determine the problem?

N/A since this feature does not come with an SLO.

## Implementation History

- 1.36: first KEP revision and implementation

## Drawbacks

It will have limited usefulness once scoring is implemented for DRA, but it is
a complicated feature and we don't have any design or timeline for when that
will happen.

## Alternatives

> ### Lexicograph ordering of ResourceSlices and resource pools based on names

Reviewer: SGTM!

ResourceSlices might also include network-atteched devices, that needs to be attached to the
node after a pod has been scheduled. This can lead to slower startup time, so typically it
will be better to allocate local devices rather than network-attached devices if possible.
Since ResourceSlices with network-attached devices will typically be published by a
controller rather than by a driver running on a node, they will be published as part of a
different resource pool than node-local devices.

### Goals

* Give DRA drivers a way to signal to the allocator the order in which devices should
  be evaluated.

### Non-Goals

* Provide a way to order the way the allocator evaluates devices across drivers.
* Full implementation of scoring.

## Proposal

### Risks and Mitigations

* The need for this feature is partly a result of not having scoring, and as a result,
  the scheduler only does a first-fit search for devices. With scoring, the order of the
  devices across resource pools and ResourceSlices doesn't matter since the allocator
  would evaluate all sets of devices to choose the one with the best fit. So this
  feature will be less relevant once we get scoring, but we don't

>   currently have any timeline for when that might happen.

Reviewer: s/when/when - or if - /

Even with scoring, we may not evaluate *every* possibility, as that may take too long. In that case, influencing the scheduler to pick devices that are likely to score higher, first, can improve scheduling latency.

* This will require that the order of ResourceSlices and resource pools are sorted
  in the allocator. This means additional work in the allocator so it might
  negatively impact the scheduling throughput for DRA workloads.

## Design Details

The proposal is to add two new fields to the ResourceSlice API, one for specifying
the priority of a ResourceSlice within a resource pool and another one to specify
the priority of resource pool relative to other resource pools published by the
same driver.

### API Changes

```go
type ResourceSliceSpec struct {
    ...

    // Priority specifies the order in which the allocator should
    // evaluate devices from ResourceSlices when allocating devices
    // for ResourceClaims. A higher value means the allocator will
    // evaluate devices from the ResourceSlice before devices from
    // ResourceSlices with a lower value. Devices from ResourceSlices
    // with the same value can be evaluated in any order.
    //
    // This field is optional and ResourceSlices without a value
    // specified is assumed to have a priority of zero. This means

>     // that a negative value means that devices in the ResourceSlice

Reviewer: ```suggestion
    // that devices in a ResourceSlice with a negative value
```

    // resource pools with a lower value. Devices from resource pools
    // with the same value can be evaluated in any order. The ordering
    // only applies to resource pools from the same driver. DRA does
    // not provide a way to specify the ordering between resource pools
    // from different drivers.
    //
    // This field is optional and resource pool without a value
    // specified is assumed to have a priority of zero. This means
    // that a negative value means that devices in the resource pool
    // will be evaluated after devices in resource pools without a
    // value specified.
    //
    // +optional
    Priority *int64
}
```

### Implementation Details

The priority fields will be used in the `GatherPools` function in
the [allocator package](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/dynamic-resource-allocation/structured/internal/experimental/pools_experimental.go#L58).
The function will be updated to sort the ResourceSlices within each
resource pool based on the `spec.priority` field and the resource pools
within each driver will be sorted based on the `spec.pool.priority` field.

### Validation

> The `spec.pool.priority` field will be set on all ResourceSlices within the

Author: That might an option. But the proposal for the `ResourcePool` object in https://github.com/kubernetes/enhancements/issues/5677 is just eventually consistent information about information, so putting information like the priority in there will have broader consequences. At least it means the scheduler will need to be aware of it and it will need to be published by the drivers along with the ResourceSlices. But avoiding duplicate information across ResourceSlices would be useful.
